libsplinter/src/node_registry/yaml/remote.rs

```diff
@@ -0,0 +1,337 @@
+                                 forced_refresh_period may be too large",
+                            )
+                        })
+                    })
+                    .transpose()?;
+                Ok(())
+            })
+    }
+
+    /// Attempt to refresh the internal cache if necessary and return the cache's contents.
+    fn get_nodes(&mut self) -> Result<Vec<Node>, NodeRegistryError> {
+        // If the last attempt to refresh the cache wasn't successful, try again
+        if !self.last_refresh_successful {
+            match self.refresh_cache() {
+                Ok(_) => debug!("Successfully refreshed remote registy '{}'", self.url),
+                // Last attempt also failed, so just log with DEBUG to keep the WARN logs clean
+                Err(err) => debug!("Failed to refresh remote registry '{}': {}", self.url, err),
+            }
+        }
+        // If the forced refresh period has elapsed, attempt to refresh the cache
+        else if self
+            .next_forced_refresh
+            .map(|instant| instant < Instant::now())
+            .unwrap_or(false)
+        {
+            match self.refresh_cache() {
+                Ok(_) => debug!("Forced refresh of remote registy '{}' successful", self.url),
+                // Already checked that the previous attempt was successful (previous branch of the
+                // if/else), so log as WARN to indicate that something's changed
+                Err(err) => warn!(
+                    "Forced refresh of remote registry '{}' failed: {}",
+                    self.url, err
+                ),
+            }
+        }
+
+        self.cache.get_cached_nodes()
+    }
+}
+
+/// Fetch, parse, and validate the YAML node registry file at the given URL.
+fn fetch_nodes_from_remote(url: &str) -> Result<Vec<Node>, NodeRegistryError> {
+    let bytes = reqwest::blocking::get(url)
+        .and_then(|response| response.error_for_status())
+        .map_err(|err| {
+            NodeRegistryError::general_error_with_source(
+                &format!("Failed to fetch remote registry file from {}", url),
+                Box::new(err),
+            )
+        })?
+        .bytes()
+        .map_err(|err| {
+            NodeRegistryError::general_error_with_source(
+                "Failed to get bytes from remote registry file HTTP response",
+                Box::new(err),
+            )
+        })?;
+    let nodes: Vec<Node> = serde_yaml::from_slice(&bytes).map_err(|_| {
+        NodeRegistryError::general_error(
+            "Failed to deserialize remote registry file: Not a valid YAML sequence of nodes",
+        )
+    })?;
+
+    validate_nodes(&nodes)?;
+
+    Ok(nodes)
+}
+
+/// Ifinitely loop, attempting to refresh the `internal` cache every `refresh_period`.
+fn automatic_refresh_loop(refresh_period: Duration, internal: Arc<Mutex<Internal>>, url: &str) {
+    loop {
+        thread::sleep(refresh_period);
```

Comment:
This seems like a place to use some kind of a atomic bool to signal shutdown of this thread.

Reply:
Updated to use an AtomicBool to signal shutdown